File: src/main/java/aurocosh/divinefavor/common/item/ItemWarpMarker.java
package aurocosh.divinefavor.common.item;

import aurocosh.divinefavor.common.core.DivineFavorCreativeTab;
import aurocosh.divinefavor.common.item.base.ModItem;
import aurocosh.divinefavor.common.util.UtilEntity;
import aurocosh.divinefavor.common.util.UtilNbt;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.item.ItemStack;
import net.minecraft.nbt.NBTTagCompound;
import net.minecraft.util.ActionResult;
import net.minecraft.util.EnumActionResult;
import net.minecraft.util.EnumFacing;
import net.minecraft.util.EnumHand;
import net.minecraft.util.math.BlockPos;
import net.minecraft.world.World;

public class ItemWarpMarker extends ModItem {
    public static String TAG_POSITION = "Position";
    public static String TAG_DIMENSION = "Dimension";

    private boolean canTeleportToDimensions;

    public ItemWarpMarker(String name, boolean canTeleportToDimensions) {
        super(name, name);
        this.canTeleportToDimensions = canTeleportToDimensions;
        setMaxStackSize(1);
        setCreativeTab(DivineFavorCreativeTab.INSTANCE);
    }

    @Override
    public ActionResult<ItemStack> onItemRightClick(World world, EntityPlayer player, EnumHand hand) {
        ItemStack stack = player.getHeldItem(hand);
        if(!teleportPlayer(stack, player, world))
            return new ActionResult<>(EnumActionResult.PASS, stack);
        return new ActionResult<>(EnumActionResult.SUCCESS, ItemStack.EMPTY);
    }

    @Override
    public EnumActionResult onItemUse(EntityPlayer player, World world, BlockPos pos, EnumHand hand, EnumFacing facing, float hitX, float hitY, float hitZ) {
        ItemStack stack = player.getHeldItem(hand);
        if(!teleportPlayer(stack, player, world))
            return EnumActionResult.PASS;
        return EnumActionResult.SUCCESS;
    }

    private boolean teleportPlayer(ItemStack stack, EntityPlayer player, World world){
        if(world.isRemote)
            return false;
        if(stack.isEmpty())
            return false;
        if(!(stack.getItem() instanceof ItemWarpMarker))
            return false;
        NBTTagCompound nbt = UtilNbt.getEistingOrNewNBT(stack);
        if(!UtilNbt.checkForTags(nbt, TAG_POSITION, TAG_DIMENSION))
            return false;

        BlockPos destination = UtilNbt.getBlockPos(nbt, TAG_POSITION);
        int dimension = nbt.getInteger(TAG_DIMENSION);
        if(!canTeleportToDimensions && dimension != player.dimension)
            return false;

        UtilEntity.teleport(player, dimension, destination);

        stack.shrink(1);
        if (stack.isEmpty())
            player.inventory.deleteStack(stack);
        return true;
    }
}

File: src/main/kotlin/aurocosh/divinefavor/common/item/ItemWarpMarker.kt
package aurocosh.divinefavor.common.item

import aurocosh.divinefavor.DivineFavor
import aurocosh.divinefavor.common.constants.ConstGemTabOrder
import aurocosh.divinefavor.common.item.base.ModItem
import aurocosh.divinefavor.common.lib.extensions.compound
import aurocosh.divinefavor.common.lib.extensions.getBlockPos
import aurocosh.divinefavor.common.lib.extensions.hasKey
import aurocosh.divinefavor.common.util.UtilEntity
import net.minecraft.entity.player.EntityPlayer
import net.minecraft.item.ItemStack
import net.minecraft.util.ActionResult
import net.minecraft.util.EnumActionResult
import net.minecraft.util.EnumFacing
import net.minecraft.util.EnumHand
import net.minecraft.util.math.BlockPos
import net.minecraft.world.World

class ItemWarpMarker(name: String, private val canTeleportToDimensions: Boolean) : ModItem(name, name, ConstGemTabOrder.OTHER_GEMS) {

    init {
        setMaxStackSize(64)
        creativeTab = DivineFavor.TAB_GEMS
    }

    override fun onItemRightClick(world: World, player: EntityPlayer, hand: EnumHand): ActionResult<ItemStack> {
        val stack = player.getHeldItem(hand)
        return if (!teleportPlayer(stack, player, world)) ActionResult(EnumActionResult.PASS, stack) else ActionResult(EnumActionResult.SUCCESS, stack)
    }

    override fun onItemUse(player: EntityPlayer, world: World, pos: BlockPos, hand: EnumHand, facing: EnumFacing, hitX: Float, hitY: Float, hitZ: Float): EnumActionResult {
        val stack = player.getHeldItem(hand)
        return if (!teleportPlayer(stack, player, world)) EnumActionResult.PASS else EnumActionResult.SUCCESS
    }

    private fun teleportPlayer(stack: ItemStack, player: EntityPlayer, world: World): Boolean {
        if (world.isRemote)
            return false
        if (stack.isEmpty)
            return false
        if (stack.item !is ItemWarpMarker)
            return false
        val tag = stack.compound
        if (!tag.hasKey(TAG_POSITION, TAG_DIMENSION))
            return false

        val destination = tag.getBlockPos(TAG_POSITION)
        val dimension = tag.getInteger(TAG_DIMENSION)
        if (!canTeleportToDimensions && dimension != player.dimension)
            return false

        UtilEntity.teleport(player, dimension, destination)
        stack.shrink(1)
        return true
    }

    companion object {
        const val TAG_POSITION = "Position"
        const val TAG_DIMENSION = "Dimension"
    }
}